[Bundle 5/n] Prune g_chainman usage in RPC modules

dongcarl opened the pull request:
Overall PR: #20158 (tree-wide: De-globalize ChainstateManager)

Based on:
- [x] #21270 | [Bundle 4/n] Prune g_chainman usage in validation-adjacent modules
- [x] #21525 | [Bundle 4.5/n] Followup fixups to bundle 4

Note to reviewers:
1. This bundle may _apparently_ introduce usage of `g_chainman` or `::Chain(state|)Active()` globals, but these are resolved later on in the overall PR. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
2. There may be seemingly obvious local references to `ChainstateManager` or other validation objects which are not being used in callers of the current function in question, this is done intentionally to **_keep each commit centered around one function/method_** to ease review and to make the overall change systematic. We don't assume anything about our callers. Rest assured that once we are considering that particular caller in later commits, we will use the obvious local references. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
3. When changing a function/method that has many callers (e.g. `LookupBlockIndex` with 55 callers), it is sometimes easier (and less error-prone) to use a scripted-diff. When doing so, there will be 3 commits in sequence so that every commit compiles like so:
	1. Add `new_function`, make `old_function` a wrapper of `new_function`, divert all calls to `old_function` to `new_function` **in the local module only**
	2. Scripted-diff to divert all calls to `old_function` to `new_function` **in the rest of the codebase**
	3. Remove `old_function`

dongcarl committed:
rest: Pass in NodeContext to rest_block

dongcarl committed:
rpc/blockchain: Use existing NodeContext

Also pass in appropriate object to:
- BIP9SoftForkDescPushBack
- BuriedForkDescPushBack

dongcarl committed:
rpc/mining: Use existing NodeContext

Also pass in appropriate object to:
- GetNetworkHashPS
- [gG]enerateBlock{,s}

Also:
- Misc style/constness changes

dongcarl committed:
rpc/rawtx: Use existing NodeContext

Also pass in appropriate object to:
- TxToJSON

dongcarl committed:
rest: Use existing NodeContext

dongcarl committed:
rpc/*,rest: Add review-only assertion to EnsureChainman

ryanofsky reviewed (approved):
Code review ACK last 6 commits only a041676d3905002758a9c1827b410ea87d3329cf. Many chainmans.

- [X] 384c56d16e724344f092192d31ef01a529f1bfec rpc/*,rest: Add review-only assertion to EnsureChainman (14/19)
- [X] b6d384784f9cf272f1a56600a770faf8d2d8f05b rpc/blockchain: Use existing NodeContext (15/19)
- [X] 2b0fb0e2219f3ffe5dee1e45ad519ff427fa01b6 rpc/mining: Use existing NodeContext (16/19)
- [X] 8e3d4558a1d6ea55279642e4a9279d4b4c31ed69 rpc/rawtx: Use existing NodeContext (17/19)
- [X] cbe4f7ba2165518800ffd69757b5ec79fac9a7c9 rest: Pass in NodeContext to rest_block (18/19)
- [X] a041676d3905002758a9c1827b410ea87d3329cf rest: Use existing NodeContext (19/19)

dongcarl commented:
Pushed a041676d3905002758a9c1827b410ea87d3329cf -> 6b89f537bad5fc3761ef3ada4cc5c9ac213c0ae6
- Fixed assert linter complaint by using `CHECK_NONFATAL` instead of `assert`

MarcoFalke commented:
(Needs rebase)

dongcarl commented:
Pushed 6b89f537bad5fc3761ef3ada4cc5c9ac213c0ae6 -> b57dc39eae25294071959b1eec42bd7dd5563357
- Rebased over master
- Prepended fixes to last bundle

-----

Note to reviewers:
- I've made some changes in 5de3e3609ab591d88281558182efdec39b065541 that requires some attention, mostly I had to make decisions about default values of `getNumBlocks`, `isInitialBlockDownload`, `getUnspentOutput`, and `checkFinalTx` when our context doesn't contain a chainman
- The commits before b7ed70dba191881d79fe71a746a2a826f40de884 should address the following posthumous reviews:
  - From jnewbery:
    - https://github.com/bitcoin/bitcoin/pull/21270#issuecomment-796738048
      - I didn't fix this one, but I added a `TODO` comment so that we don't lost track of it
    - https://github.com/bitcoin/bitcoin/pull/21270#discussion_r592291225
    - https://github.com/bitcoin/bitcoin/pull/21270#discussion_r592296942
    - https://github.com/bitcoin/bitcoin/pull/21270#discussion_r592299738
    - https://github.com/bitcoin/bitcoin/pull/21270#discussion_r592301704
  - From MarcoFalke:
    - https://github.com/bitcoin/bitcoin/pull/21270#discussion_r593096212
    - https://github.com/bitcoin/bitcoin/pull/21270#discussion_r593097032
    - https://github.com/bitcoin/bitcoin/pull/21270#discussion_r593097867
    - https://github.com/bitcoin/bitcoin/pull/21270#discussion_r593100570

I have also went through all my changes in the last bundle and ensured that we didn't miss any potential UB-introductions.

A few explanations:

1. https://github.com/bitcoin/bitcoin/pull/21270#discussion_r592293190
Unfortunately GetUTXOStats is fuzzed by constructing a `CCoinsView*` as an input, so this signature needs to remain.

2. https://github.com/bitcoin/bitcoin/pull/21270#discussion_r592299232
100% agree that it's unnecessary, but going to leave it be for now since it'll be removed in the end if that's okay with folks.

3. https://github.com/bitcoin/bitcoin/pull/21270#discussion_r592357247
Right, it would most likely be better to take a local `const CChainstate&` once in places where we're already taking `cs_main`, let's do this in a followup!

jnewbery commented:
Concept ACK. Thanks for addressing review comments from the last bundle!

MarcoFalke commented on src/miner.h:
```diff
@@ -146,6 +146,7 @@ class BlockAssembler
     int64_t nLockTimeCutoff;
     const CChainParams& chainparams;
     const CTxMemPool& m_mempool;
+    CChainState& m_chainstate;
```
nit: Can this be const?

dongcarl replied:
Unfortunately not... We pass `m_chainstate` to `TestBlockValidity` in `BlockAssembler::CreateNewBlock`, and `TestBlockValidity` requires its `CChainState` param to be non-const

MarcoFalke commented on src/node/interfaces.cpp:
```diff
@@ -621,8 +651,13 @@ class ChainImpl : public Chain
-        assert(std::addressof(::ChainstateActive()) == std::addressof(m_node.chainman->ActiveChainstate()));
-        return m_node.chainman->ActiveChainstate().IsInitialBlockDownload();
+        const CChainState* active_chainstate;
+        if (m_node.chainman) {
```
does it make sense to run a node without a chainman? Imo it might make more sense to use Assert, like it is already done in `waitForNotificationsIfTipChanged`.

Maybe the assert can even be wrapped into a private member:

```
private:
 ChainstateManager& chainman() { return *Assert(m_node.chainman); }

MarcoFalke commented on src/node/transaction.cpp:
```diff
@@ -39,6 +39,7 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
 
     { // cs_main scope
     LOCK(cs_main);
+    assert(node.chainman);
```
would be nice to assert before cs_main, because cs_main "belongs" to the chainman

MarcoFalke commented on src/node/coin.cpp:
```diff
@@ -12,6 +12,7 @@ void FindCoins(const NodeContext& node, std::map<COutPoint, Coin>& coins)
 {
     assert(node.mempool);
     LOCK2(cs_main, node.mempool->cs);
+    assert(node.chainman);
```
would be nice to assert before cs_main, because cs_main "belongs" to the chainman

maflcko reviewed (commented):
left some questions

jnewbery reviewed (commented):
Looks good. I've left a few style comments inline.

This is quite a large PR. I wonder if this should perhaps be split into two PRs - one to address the comments from bundle 4 and another for bundle 5. I don't think they overlap very much so you can have them open in parallel without much conflict.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -182,38 +182,51 @@ class NodeImpl : public Node
-        const CBlockIndex* tip = WITH_LOCK(::cs_main, return m_context->chainman->ActiveChain().Tip());
+        const CBlockIndex* tip;
+        if (m_context->chainman) {
+            LOCK(cs_main);
```
In commit "node/interfaces: Meaningful locking and avoid potential UB" (5de3e3609ab591d88281558182efdec39b065541)

It seems like this method and the two IBD methods are the only places that are changing locks or adding "meaningful locking". Since locking changes can have unexpected pitfalls I think it'd be good to split the locking parts of this commit into a separate commit. Also would be good to expand the commit message to say what's motivating the lock changes. I'm assuming it's to satisfy some lock annotations later?

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -182,38 +182,51 @@ class NodeImpl : public Node
     {
-        assert(std::addressof(::ChainActive()) == std::addressof(m_context->chainman->ActiveChain()));
-        const CBlockIndex* tip = WITH_LOCK(::cs_main, return m_context->chainman->ActiveChain().Tip());
+        const CBlockIndex* tip;
+        if (m_context->chainman) {
```
In commit "node/interfaces: Meaningful locking and avoid potential UB" (5de3e3609ab591d88281558182efdec39b065541)

> Same here. You don't need this local variable if you exit early when `chainman` doesn't exist.

I think John's suggestions are good, but Marco's suggestion to use an accessor that asserts https://github.com/bitcoin/bitcoin/pull/21391#discussion_r596601383 would be even better. That way we don't need to waste brain cells thinking about what values to return if chainman is null. Like if getBestBlockHash returns the genesis block hash in that case should getNumBlocks return 1 instead of 0 to be consistent? Easier to take everything down and assert.

jnewbery replied:
Seems reasonable to me!

In general, I like the concept of making all components optional, since it enforces decoupling. However, it's hard to imagine what a bitcoind without chainman would look like.

ryanofsky reviewed (approved):
Code review ACK b57dc39eae25294071959b1eec42bd7dd5563357. Looks good. I just rereviewed from scratch instead of seeing what changed since the last review. Reviewing with `-U0 --word-diff-regex=.` made it easy to check that the bulk of this PR is just switching from one chainman reference to another, and to focus more closely on the parts making other changes. `-w --color-moved` was also useful for reviewing the "Move comments to declarations" commit

dongcarl commented:
Pushed b57dc39eae25294071959b1eec42bd7dd5563357 -> 877c8a01fc1137e0c8187b0b6990a99658077152
- Addressed some review suggestions
- Note: this is not yet rebased

dongcarl commented:
Pushed 877c8a01fc1137e0c8187b0b6990a99658077152 -> b7833034a554b571b3d3e3e55a0639421770dfb9
- Rebased on top of #21525
- Addressed all reviews

dongcarl commented:
Pushed b7833034a554b571b3d3e3e55a0639421770dfb9 -> 677b836c4847315bf24a9d222bbb2eb23a2b9a2c
- Use `CHECK_NONFATAL` instead of assert

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -183,21 +185,25 @@ class NodeImpl : public Node
+        const CBlockIndex* tip;
+        {
+            LOCK(cs_main);
+            assert(std::addressof(::ChainActive()) == std::addressof(chainman().ActiveChain()));
```
In commits "node/ifaces: NodeImpl: Use an accessor for ChainMan" (3460a07d88d841c009e7b7efc0be2ae89bb4a63e)
and "node/ifaces: ChainImpl: Use an accessor for ChainMan" (3b75ec491a68eb6fb137c2f829e7dde93b3c3189)

Not important unless there will be a lot more of these, but I think it would be better not to add nested scopes for temporary asserts that we are going to want to remove and collapse later. The asserts seem good in other places where they can be removed with scripted diffs just deleting the assert lines. But places like this should be easier to clean up later using a helper like:

```c++
//! Temporary accessor that can be removed after
//! https://github.com/bitcoin/bitcoin/pull/20158, but helps check correctness
//! of refactored code in the meantime.
static inline CChain& ActiveChain(ChainstateManager& chainman) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
{
    assert(std::addressof(::ChainActive()) == std::addressof(chainman.ActiveChain()));
    return chainman.ActiveChain();
}
```

that could substituted out with a scripted-diff without needing to change surrounding code. Could also be appealing if you think usage looks nicer:

```diff
-        assert(std::addressof(::ChainActive()) == std::addressof(chainman().ActiveChain()));
-        return chainman().ActiveChain().Height();
+        return ActiveChain(chainman()).Height();
```

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -475,19 +475,71 @@ class PeerManagerImpl final : public PeerManager
+     * May disconnect from the peer in the case of a bad request.
+     *
+     * @param[in]   peer            The peer that we received the request from
+     * @param[in]   chain_params    Chain parameters
```
In commit "net_processing: Move comments to declarations" (067d0ba36cb701adff530f76ef9b1f574c966c6f)

Is no more chain parameters.

ryanofsky reviewed (approved):
Code review ACK 677b836c4847315bf24a9d222bbb2eb23a2b9a2c

Since last review there were many suggested review changes, notably replacing interfaces update commit 5de3e3609ab591d88281558182efdec39b065541 with more limited 3460a07d88d841c009e7b7efc0be2ae89bb4a63e 3b75ec491a68eb6fb137c2f829e7dde93b3c3189 commits for https://github.com/bitcoin/bitcoin/pull/21391#discussion_r597546794

Cirrus android failure https://cirrus-ci.com/task/5300781294288896 is spurious and should be fixed by #21541

Left some new suggestions but they aren't important and could be followed up later or not at all.

dongcarl commented:
Pushed 677b836c4847315bf24a9d222bbb2eb23a2b9a2c -> 188bb5e224d757d8d0603bb05a911f816776d295
- Rebased on updated #21525, update summary reproduced below:

> Pushed 3b75ec491a68eb6fb137c2f829e7dde93b3c3189 -> 693414d27181cf967f787a2ca72344e52c58c7f0
> - Addressed https://github.com/bitcoin/bitcoin/pull/21391#pullrequestreview-624405217

ryanofsky reviewed (approved):
Code review ACK 188bb5e224d757d8d0603bb05a911f816776d295. Only changes since last review were comment changes following up the last review (thanks!)

MarcoFalke commented:
Needs rebase?

dongcarl commented:
Pushed 188bb5e224d757d8d0603bb05a911f816776d295 -> 7045f5f557f910827f13501e303ad617a8f0aaa6
- Rebased over master after merge of #21525, fixed silent conflict in `src/rest.cpp`

dongcarl commented:
@MarcoFalke re: https://github.com/bitcoin/bitcoin/pull/21525/files/e62067e7bcad5a559899afff2e4a8e8b7e9f4301..7b8e976cd5ac78a22f1be2b2fed8562c693af5d9#diff-cd34d7ba23b9b4f3f61cd06e9289c4981a5a6bcb17b61b5d010f3dc01c77787e

This is more or less what I'm thinking, let me know if I'm missing something!
```diff
diff --git a/src/miner.cpp b/src/miner.cpp
index 8a9406f810..3bc7fdd458 100644
--- a/src/miner.cpp
+++ b/src/miner.cpp
@@ -39,13 +39,14 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
     return nNewTime - nOldTime;
 }
 
-void RegenerateCommitments(CBlock& block, BlockManager& blockman)
+void RegenerateCommitments(CBlock& block, CBlockIndex* prev_block)
 {
     CMutableTransaction tx{*block.vtx.at(0)};
     tx.vout.erase(tx.vout.begin() + GetWitnessCommitmentIndex(block));
     block.vtx.at(0) = MakeTransactionRef(tx);
 
-    GenerateCoinbaseCommitment(block, WITH_LOCK(::cs_main, assert(std::addressof(g_chainman.m_blockman) == std::addressof(blockman)); return blockman.LookupBlockIndex(block.hashPrevBlock)), Params().GetConsensus());
+    WITH_LOCK(::cs_main, assert(g_chainman.m_blockman.LookupBlockIndex(block.hashPrevBlock) == prev_block));
+    GenerateCoinbaseCommitment(block, prev_block, Params().GetConsensus());
 
     block.hashMerkleRoot = BlockMerkleRoot(block);
 }
diff --git a/src/miner.h b/src/miner.h
index c400c90f6c..becf362b79 100644
--- a/src/miner.h
+++ b/src/miner.h
@@ -202,8 +202,7 @@ private:
 void IncrementExtraNonce(CBlock* pblock, const CBlockIndex* pindexPrev, unsigned int& nExtraNonce);
 int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParams, const CBlockIndex* pindexPrev);
 
-// TODO just accept a CBlockIndex*
 /** Update an old GenerateCoinbaseCommitment from CreateNewBlock after the block txs have changed */
-void RegenerateCommitments(CBlock& block, BlockManager& blockman);
+void RegenerateCommitments(CBlock& block, CBlockIndex* prev_block);
 
 #endif // BITCOIN_MINER_H
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index 8b54d07b50..5136a4e3de 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -374,8 +374,8 @@ static RPCHelpMan generateblock()
 
     // Add transactions
     block.vtx.insert(block.vtx.end(), txs.begin(), txs.end());
-    WITH_LOCK(::cs_main, CHECK_NONFATAL(std::addressof(g_chainman.m_blockman) == std::addressof(chainman.m_blockman)));
-    RegenerateCommitments(block, WITH_LOCK(::cs_main, return std::ref(chainman.m_blockman)));
+    CBlockIndex* prev_block = WITH_LOCK(::cs_main, return chainman.m_blockman.LookupBlockIndex(block.hashPrevBlock));
+    RegenerateCommitments(block, prev_block);
 
     {
         LOCK(cs_main);
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index f7800aefca..46c7b781b0 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -253,7 +253,8 @@ CBlock TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransa
     for (const CMutableTransaction& tx : txns) {
         block.vtx.push_back(MakeTransactionRef(tx));
     }
-    RegenerateCommitments(block, WITH_LOCK(::cs_main, return std::ref(g_chainman.m_blockman)));
+    CBlockIndex* prev_block = WITH_LOCK(::cs_main, return g_chainman.m_blockman.LookupBlockIndex(block.hashPrevBlock));
+    RegenerateCommitments(block, prev_block);
 
     while (!CheckProofOfWork(block.GetHash(), block.nBits, chainparams.GetConsensus())) ++block.nNonce;
 

```

MarcoFalke commented on src/rest.cpp:
```diff
@@ -181,13 +181,14 @@ static bool rest_headers(const std::any& context,
-        tip = ::ChainActive().Tip();
-        const CBlockIndex* pindex = g_chainman.m_blockman.LookupBlockIndex(hash);
-        while (pindex != nullptr && ::ChainActive().Contains(pindex)) {
+        ChainstateManager& chainman = EnsureChainman(context);
```
nit: Would be nice to do this before locking cs_main, because I think cs_main should be made a (private) member of chainman in the future, if possible.

MarcoFalke commented on src/rest.cpp:
```diff
@@ -249,8 +251,9 @@ static bool rest_block(HTTPRequest* req,
         LOCK(cs_main);
-        tip = ::ChainActive().Tip();
-        pblockindex = g_chainman.m_blockman.LookupBlockIndex(hash);
+        ChainstateManager& chainman = EnsureChainman(context);
```
same

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -822,8 +824,9 @@ static RPCHelpMan getblockheader()
         LOCK(cs_main);
-        pblockindex = g_chainman.m_blockman.LookupBlockIndex(hash);
-        tip = ::ChainActive().Tip();
+        ChainstateManager& chainman = EnsureChainman(request.context);
```
same

maflcko reviewed (approved):
As you are modifying those lines, it would be nice if the chainman reference was created before taking cs_main, as that would simplify future work to remove cs_main (or reduce the scope of it).

dongcarl committed:
miner: Pass in previous CBlockIndex to RegenerateCommitments

dongcarl commented:
> As you are modifying those lines, it would be nice if the chainman reference was created before taking cs_main, as that would simplify future work to remove cs_main (or reduce the scope of it).

These changes seem simple enough, so I'd be happy to make it. I do want to understand: how do you see these small changes helping with "future work to remove cs_main (or reduce the scope of it)"?

MarcoFalke commented:
cs_main is a global and used to lock stuff that has nothing to do with validation (e.g. the p2p thread, disk access, orphan handling, ...). There is ongoing effort to clean those up: #19398, https://github.com/bitcoin/bitcoin/pull/21527/files#r601283910, https://github.com/bitcoin/bitcoin/pull/21598#issuecomment-813224175, ...

If cs_main is a member, the code might look like this:

```cpp
Chainstatemanager& cm = ...;
LOCK(cm.main);
```

So simply getting a reference to the chainstatemanager before taking the lock will help in that regard, so that the two lines don't have to be switched in the future.

jnewbery commented on src/rest.cpp:
```diff
@@ -181,13 +181,14 @@ static bool rest_headers(const std::any& context,
             if (headers.size() == (unsigned long)count)
                 break;
-            pindex = ::ChainActive().Next(pindex);
+            pindex = chainman.ActiveChain().Next(pindex);
```
Maybe just grab `chainman.ActiveChain()` once above and then use it for each iteration.

jnewbery commented on src/rest.cpp:
```diff
@@ -249,8 +251,9 @@ static bool rest_block(HTTPRequest* req,
-        tip = ::ChainActive().Tip();
-        pblockindex = g_chainman.m_blockman.LookupBlockIndex(hash);
+        ChainstateManager& chainman = EnsureChainman(context);
+        tip = chainman.ActiveChain().Tip();
```
```suggestion
        tip = chainman.ActiveTip();
```

jnewbery commented on src/rest.cpp:
```diff
@@ -550,12 +554,12 @@ static bool rest_getutxos(const std::any& context, HTTPRequest* req, const std::
             // use db+mempool as cache backend in case user likes to query mempool
             LOCK2(cs_main, mempool->cs);
-            CCoinsViewCache& viewChain = ::ChainstateActive().CoinsTip();
+            CCoinsViewCache& viewChain = chainman.ActiveChainstate().CoinsTip();
```
Perhaps just call `chainman.ActiveChainstate()` once in this function.

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -198,7 +199,7 @@ static RPCHelpMan getblockcount()
 {
     LOCK(cs_main);
-    return ::ChainActive().Height();
+    return EnsureChainman(request.context).ActiveChain().Height();
```
Agree about `ActiveHeight()` and similar comments below, but may be fine to leave the `LOCK(cs_main)` line as-is since I suspect we'll move to lock annotations for `ActiveChainstate()` in the near future (vs. the nested acquisition in the function definition).

jnewbery replied:
I think that'd be a move in the wrong direction. Ideally we'd remove the need to lock `cs_main` completely from the `Active*()` methods. I'm concerned that adding so many additional cs_main locks may cause performance regressions - for example, the optimization in #8007 is no longer possible since cs_main will almost always be locked before calling `IsInitialBlockDownload()`

@jamesob: Did you consider changing the active pointer to being an atomic to remove this requirement as suggested here: https://github.com/bitcoin/bitcoin/pull/19806#issuecomment-768946522?

jamesob replied:
I think you may be forgetting that long before `ChainstateManager` was introduced, [`chainActive` was protected by `cs_main`](https://github.com/bitcoin/bitcoin/blob/0.16/src/validation.h#L445). So unless we dramatically rework the chainstate locking scheme, the `Active{Height,Tip,..}()` methods will always require holding cs_main one way or another. I think ultimately annotations are a cleaner way of doing this because they give callers the flexibility to reuse existing cs_main acquisitions, and avoid recursive locking.

Your suggestion to move chainman's active pointer to an atomic may be a good one, but I think it's largely incidental in terms of performance since we usually need to acquire cs_main anyway when referencing chainstate data. Though my comment above is unrelated to that, since I'm making reference to the chain's protection by cs_main and not ancillary ChainstateManager layout.

[Recent bitcoinperf benchmarks](https://bitcoinperf.com/d/YiV16Vsik/overview?orgId=1&from=now-1y&to=now) don't indicate any regressions in recent revisions, but I will be doing more thorough benchmarking in a bit.

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -823,8 +824,9 @@ static RPCHelpMan getblockheader()
-        tip = ::ChainActive().Tip();
+        ChainstateManager& chainman = EnsureChainman(request.context);
+        pblockindex = chainman.m_blockman.LookupBlockIndex(hash);
+        tip = chainman.ActiveChain().Tip();
```
`ActiveTip()`

jamesob commented on src/rpc/mining.cpp:
```diff
@@ -420,12 +426,13 @@ static RPCHelpMan getmininginfo()
 
     UniValue obj(UniValue::VOBJ);
-    obj.pushKV("blocks",           (int)::ChainActive().Height());
+    obj.pushKV("blocks",           (int)active_chain.Height());
```
https://github.com/bitcoin/bitcoin/pull/21391/commits/626aa73ae48f6b5a50e88fc3d03e8c7842f75b1b

Can remove unnecessary cast.

jamesob commented on src/rest.cpp:
```diff
@@ -181,13 +181,14 @@ static bool rest_headers(const std::any& context,
-        const CBlockIndex* pindex = g_chainman.m_blockman.LookupBlockIndex(hash);
-        while (pindex != nullptr && ::ChainActive().Contains(pindex)) {
+        ChainstateManager& chainman = EnsureChainman(context);
+        tip = chainman.ActiveChain().Tip();
```
`ActiveTip()`

jamesob reviewed (commented):
Looks good! Would be good to fix up the `Active*()` method usages. I think given there is no scripted-diff usage, it may also be worth squashing the last two commits together.

ryanofsky reviewed (approved):
Code review ACK 7045f5f557f910827f13501e303ad617a8f0aaa6. No changes since last review other than rebase. Cleanup suggestions from Marco/John/James all seem reasonable too.

dongcarl commented:
Question: Is there an actual difference between `chainman.ActiveTip/Height()` and `chainman.ActiveChain().Tip/height()` or is it just shorter?

jnewbery commented:
> Is there an actual difference between chainman.ActiveTip/Height() and chainman.ActiveChain().Tip/height() or is it just shorter?

Just shorter. Feel free to ignore those comments if you disagree that it's clearer.

dongcarl committed:
rpc: Add alt Ensure* functions acepting NodeContext

dongcarl committed:
rpc: Tidy up local references (see commit message)

Organize local variables/references such that:

1. There is always a `ChainstateManager` reference before any `LOCK(cs_main)`.
2. NodeContext references are used with Ensure*() functions introduced in
   previous commit where appropriate to avoid duplicate assertions.

dongcarl committed:
rpc: Remove unnecessary casting of block height

dongcarl committed:
rpc: style: Improve BuriedForkDescPushBack signature

dongcarl committed:
rpc/rest: Take and reuse local Chain/ChainState obj

In all rest/rpc-related modules, if there are multiple calls to
ActiveChain{,State}(), and the calls fall under the same ::cs_main lock,
we can simply take a local reference and use/reuse it instead of calling
ActiveChain{,State}() again and again.

dongcarl commented:
Pushed 7045f5f557f910827f13501e303ad617a8f0aaa6 -> 1c483e8e73c9f02674895b73b017dc0d13a030b0
- Addressed MarcoFalke's `cs_main` vs `ChainstateManager` code ordering comments
- Addressed jnewbery and jamesob's height int casting comments
- Addressed jnewbery's BuriedForkDescPushBack signature comments

dongcarl commented:
Pushed 1c483e8e73c9f02674895b73b017dc0d13a030b0 -> 6d923137f6eb31ace6f2d38480708af039a21437
- Once more, with feeling...
- I renamed params in last commit but forgot to rename references... oops

ryanofsky commented on src/rpc/blockchain.h:
```diff
@@ -57,8 +57,11 @@ void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fInclud
 void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr);
 
 NodeContext& EnsureNodeContext(const std::any& context);
+CTxMemPool& EnsureMemPool(const NodeContext& node);
 CTxMemPool& EnsureMemPool(const std::any& context);
+ChainstateManager& EnsureChainman(const NodeContext& node);
 ChainstateManager& EnsureChainman(const std::any& context);
+CBlockPolicyEstimator& EnsureFeeEstimator(const NodeContext& node);
 CBlockPolicyEstimator& EnsureFeeEstimator(const std::any& context);
```
In commit "rpc: Add alt Ensure* functions acepting NodeContext" (306b1cd3eeb2502904ed4698646d2c86d028aad2)

Caramba. I think this commit is ok, but it is a footgun to have overloaded functions where one of the functions takes an untyped std::any argument, because someone could easily try to call the typed overload, and fail because the type is a little off (const instead of nonconst, pointer instead of reference, or just accidentally passing the wrong variable), and the compiler will never catch it because it will fall back to the std::any overload.

I think the new functions are good, but the existing std::any functions should be renamed to EnsureAnyMemPool, EnsureAnyChainman, etc to be more distinct. Or maybe the std::any functions should take JSONRPCRequest arguments instead. Or maybe they can be removed after a lot of usages are removed in the next commit.

Anyway no need to worry about now. Save for bundle √26

ryanofsky reviewed (approved):
Code review ACK 6d923137f6eb31ace6f2d38480708af039a21437. No changes to existing commits, just new commits doing suggested cleanups, particularly making RPC code more consistent about how it accesses chainmem & mempool globals. I like the new commit message that says "(see commit message)". Seems worth a try, people never read commit messages!

dongcarl committed:
scripted-diff: rest/rpc: Use renamed EnsureAny*()

-BEGIN VERIFY SCRIPT-
sed -i -E 's@Ensure([^(]+)(\((request\.|)context\))@EnsureAny\1\2@g' \
    -- src/rest.cpp src/rpc/*.cpp
-END VERIFY SCRIPT-

dongcarl committed:
rpc: Add renamed EnsureAny*() functions

- The original Ensure*(const std::any& context) functions are kept and
  the parameter renamed to ctx so that the scripted-diff in the
  subsequent commit will work as expected

- The renaming avoids overloading mistakes arising out of the untyped
  std::any argument.

dongcarl committed:
rest/rpc: Remove now-unused old Ensure functions

The scripted-diff in the previous commit should have removed all calls
to functions like: Ensure(?!Any)\(const std::any& (context|ctx)\), so we
can remove them now.

dongcarl commented:
Pushed 6d923137f6eb31ace6f2d38480708af039a21437 -> 586190f0b4740457cb86cba632e3d64e6dfe9b0c
- Addressed https://github.com/bitcoin/bitcoin/pull/21391#discussion_r612548023 to keep irrational numbers at bay
- Addressed jnewbery's reviews relating to taking a reference once and reusing it

ryanofsky reviewed (approved):
Code review ACK 586190f0b4740457cb86cba632e3d64e6dfe9b0c. Since last review, no changes to existing commits, just some simple new commits added: three new commits renaming std::any Ensure functions (scripted diff commit and manual pre/post commits), and one new commit factoring out a repeated `ActiveChain()` call made in a loop. Thanks for the updates!

jnewbery commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1598,9 +1606,11 @@ static RPCHelpMan utxoupdatepsbt()
-        const CTxMemPool& mempool = EnsureMemPool(request.context);
+        NodeContext& node = EnsureAnyNodeContext(request.context);
+        const CTxMemPool& mempool = EnsureMemPool(node);
+        ChainstateManager& chainman = EnsureChainman(node);
```
I haven't commented on all of these examples, but I really do think readability is helped by placing the `Ensure*()` assumptions at the top of the rpc methods.

jnewbery reviewed (commented):
utACK 586190f0b4740457cb86cba632e3d64e6dfe9b0c

It's not a blocker, but I don't like the proliferation of code outside validation holding `m_chain`. In general, I don't think there's any reason for code outside validation to take a reference to a `CChain` - it should be able to call a higher abstraction function on chainman to read whatever data it needs.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -2261,15 +2308,17 @@ static RPCHelpMan scantxoutset()
-            pcursor = std::unique_ptr<CCoinsViewCursor>(::ChainstateActive().CoinsDB().Cursor());
+            CChainState& active_chainstate = chainman.ActiveChainstate();
+            active_chainstate.ForceFlushStateToDisk();
+            pcursor = std::unique_ptr<CCoinsViewCursor>(active_chainstate.CoinsDB().Cursor());
```
Shouldn't Cursor return a unique_ptr?

jnewbery replied:
Yes, you're right. `Cursor()` is passing ownership to the caller.

jnewbery replied:
I'm going to mark this comment as resolved since this PR shouldn't make any changes here. I think `Cursor()`'s return type should be changed in a separate PR.

MarcoFalke replied:
(unresolved comment to make it easier to see all that needs to be addressed in follow-ups)

MarcoFalke commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1600,7 +1603,7 @@ static RPCHelpMan utxoupdatepsbt()
         const CTxMemPool& mempool = EnsureMemPool(request.context);
         LOCK2(cs_main, mempool.cs);
-        CCoinsViewCache &viewChain = ::ChainstateActive().CoinsTip();
+        CCoinsViewCache &viewChain = EnsureChainman(request.context).ActiveChainstate().CoinsTip();
```
same ...

maflcko reviewed (commented):
I looked at d7824acdb9 🐃

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

I looked at d7824acdb9 🐃
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjV+AwAt/NhM4OI3plsZuJBMc1+PcsEelcHiTdtzMCG+qIWOyxE7zU11aZPYkJ2
xhT+gSEwHQxSWENkLb7hf5eU4l23JXvA9JD7WwO3yAptxMQwpmTPzOzyefAONC/z
2Up7r09SMieAwKKTLuHiAXTSuoA9vqx3dw361XsFhcnM1Pu06zqkYzYumduL9yP6
fSj22cFffihAaZEIg3leD2LE1CXWcaq80so3kWAv741hXZRx+vPssKUIdNDW2oNr
50pbZTaiKc7T1wU16KwxeY9xBj1x12pg4WrMhYRDAXIUzwP4gk18gmhhth9EvD6q
v0bfig+bk4bXt07HvEftTT7yWpMu381mNGZGrbljgGS4NomCRbBKfuvDNqsX17LB
z8dn54TrW5Pr3//Q46xWZLxhH9f04qH5D6nuhxoghdRXP52Zh1h7A/tRrm2tFJ1t
TNw7dlDWAUZs8NUlY1Jq6nZlm9DpueU+5/0tubX9fmqYO1k8IFI3T77btrsYC/0L
0gSquAFL
=bVUN
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `d97e47afcf29e46d1bcd8c3ce2c34513261879f42b7516ce21deecae7598dee6  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108d97e47afcf29e46d1bcd8c3ce2c34513261879f42b7516ce21deecae7598dee6f0103b038e27b9d6f22423192981e0a6b09a08fff0104c2741df0684649118f600826a03ec0d08f104607ab7d3f008bb756842fd7d5e1a0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0105d6f53a1f684cd90890dc3a5fbd10e4d08f104607ac5adf008456a2387879be05c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010db03aa411b2d84fdf40fd0e5b472f66408f104607ac5acf008c376081a37d662340083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f7a0f7bdfa0295b4a2d69e6a0b327eff08f020146476a00f9acf2bd9df56edec728e2465778c06a0162d1438783da6eac1f1c508f0204f5011dabfa2f4d4f12ef70ea9d5d96a3e8978e9ad8e572fdb48dc006d0859dd08f104607ac5adf008f0c82f5d5eef1e690083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented on src/rest.cpp:
```diff
@@ -181,7 +181,7 @@ static bool rest_headers(const std::any& context,
     headers.reserve(count);
```
6fb65b49f4: the scripted diff will modify files that are not tracked by git. Could use `...  -- $(git grep -l Ensure)`

maflcko reviewed (commented):
review ACK 586190f0b4740457cb86cba632e3d64e6dfe9b0c 🍯

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 586190f0b4740457cb86cba632e3d64e6dfe9b0c 🍯
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjIYgv/YeuLpuAboBxi/xa/PWLK/q/57E44wyfEF9cZHpOCgFzkR+gfJI1ouI1C
bBUBnk0CkOqLfFwi8a1Vzel2zXYvTOtYJ6OPpeau+bxJVhlaI8ysFRuI8oomn4Tx
ypQdNYi0bNMIIiuwezoYX+dPMkBRA8Ox8JuMZ/SUuO1ZH8F66s2U9EstshDhFiS5
cN9L9kujU0ZovHAnutbEF5zSFDqR/brrmuhCg0ACE5rCla87AAQqe/29SIcKvbls
TiV9EEHWj5V8eRy/CLg2ch+bGOFp4vISWJBITYdho7hnwMLRMTHC/K5RvsfR703R
t3reCFwd4whXkcBGN6z43f16GM3fvRNJVKKT7rQ0FB6n0oVjkFE3Gqxln5Ke209N
wDAbn99WerhV8tGhOhgYTcG1N9awQBTH4aZEuOt2MrtGxYkUEXw7+mchbrQ4iPPs
75AoZ51CaEWESzIYWRV82IBtNCtD0UUvGUOuo3E3fXCqrVKBmU7JnSuMDQ3UaPR5
QVKk2RIm
=kNPA
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `3e1b6d60dc76018210f94a83c678a1e0f9c63606e06fb1b3d34881ee98abc0d8  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401083e1b6d60dc76018210f94a83c678a1e0f9c63606e06fb1b3d34881ee98abc0d8f0105b7ec5a750258e1f9350975a6797d82e08fff010c608ce48ac596ca8dda4e3e191ebaaf708f0207eb9d4ab72b10e7cc64cef7b9320c1eafec56ab988bae188b944114db3a2984908f020601fae59f0f9ce3c6840ec7a0009848415073b039c0fda2a1d3a03ba46cd2a5e08f104607af238f00815a63c5066ebe6b40083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010cffa9e5392e31f6f975a915cd1b7e06708f104607b0013f008b35a5085e7d441680083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010d5f1b2a1005bf1fe39aea3d397c49c6e08f1200beb8e523ec9e3bcc4abf26b70d7ccc407b551d38dd845151da54251a7b255e808f02068c2eea595c47a8bb60704306857029f186222ebe11b8e9c29b760c970b4641008f104607b0012f008a2eb3efa0ded36060083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f22daf4bf6a5f7d5d2c1a18c5d3a2e4b08f120433629478d3299ba9269c68a611be696ec08d284ef4ee78e4df3ef11a5d48d6d08f104607b0013f0086495f6bbd1da0b1e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1069,13 +1092,21 @@ static RPCHelpMan gettxoutsetinfo()
+    {
+        LOCK(::cs_main);
+        coins_view = &active_chainstate.CoinsDB();
+        blockman = &active_chainstate.m_blockman;
```
This doesn't require `cs_main`.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1069,13 +1092,21 @@ static RPCHelpMan gettxoutsetinfo()
-    CCoinsView* coins_view = WITH_LOCK(::cs_main, return &::ChainstateActive().CoinsDB());
-    NodeContext& node = EnsureNodeContext(request.context);
-    if (GetUTXOStats(coins_view, WITH_LOCK(::cs_main, return std::ref(g_chainman.m_blockman)), stats, hash_type, node.rpc_interruption_point)) {
+    CCoinsView* coins_view;
+    BlockManager* blockman;
```
It's slightly confusing to take the references from `CoinsDB()` and `m_blockman` and take their addresses to make pointers, and then just dereference those pointers. Better just to use references as the local variables.
